src/metrics/spanner-metrics-exporter.ts

```diff
@@ -0,0 +1,138 @@
+import {ExportResult, ExportResultCode} from '@opentelemetry/core';
+import {ExporterOptions} from './external-types';
+import {MetricServiceClient} from '@google-cloud/monitoring';
+import {transformResourceMetricToTimeSeriesArray} from './transform';
+import {status} from '@grpc/grpc-js';
+
+// Stackdriver Monitoring v3 only accepts up to 200 TimeSeries per
+// CreateTimeSeries call.
+const MAX_BATCH_EXPORT_SIZE = 200;
+
+/**
+ * Format and sends metrics information to Google Cloud Monitoring.
+ */
+export class CloudMonitoringMetricsExporter implements PushMetricExporter {
+  private _projectId: string | void | Promise<string | void>;
+
+  private readonly _client: MetricServiceClient;
+
+  constructor({auth}: ExporterOptions) {
+    this._client = new MetricServiceClient({auth: auth});
+
+    // Start this async process as early as possible. It will be
+    // awaited on the first export because constructors are synchronous
+    this._projectId = auth.getProjectId().catch(err => {
+      console.error(err);
+    });
+  }
+
+  /**
+   * Implementation for {@link PushMetricExporter.export}.
+   * Calls the async wrapper method {@link _exportAsync} and
+   * assures no rejected promises bubble up to the caller.
+   *
+   * @param metrics Metrics to be sent to the Google Cloud Monitoring backend
+   * @param resultCallback result callback to be called on finish
+   */
+  export(
+    metrics: ResourceMetrics,
+    resultCallback: (result: ExportResult) => void,
+  ): void {
+    this._exportAsync(metrics).then(resultCallback, err => {
+      console.error(err.message);
+      resultCallback({code: ExportResultCode.FAILED, error: err});
+    });
+  }
+
+  async shutdown(): Promise<void> {}
+  async forceFlush(): Promise<void> {}
+
+  /**
+   * Asnyc wrapper for the {@link export} implementation.
+   * Writes the current values of all exported {@link MetricRecord}s
+   * to the Google Cloud Monitoring backend.
+   *
+   * @param resourceMetrics Metrics to be sent to the Google Cloud Monitoring backend
+   */
+  private async _exportAsync(
+    resourceMetrics: ResourceMetrics,
+  ): Promise<ExportResult> {
+    if (this._projectId instanceof Promise) {
+      this._projectId = await this._projectId;
+    }
+
+    if (!this._projectId) {
+      const error = new Error('expecting a non-blank ProjectID');
+      console.error(error.message);
+      return {code: ExportResultCode.FAILED, error};
+    }
+
+    const timeSeriesList =
+      transformResourceMetricToTimeSeriesArray(resourceMetrics);
+
+    let failure: {sendFailed: false} | {sendFailed: true; error: Error} = {
+      sendFailed: false,
+    };
+    await Promise.all(
+      this._partitionList(timeSeriesList, MAX_BATCH_EXPORT_SIZE).map(
+        async batchedTimeSeries => this._sendTimeSeries(batchedTimeSeries),
+      ),
+    ).catch(e => {
+      const error = e as {code: number};
+      if (error.code === status.PERMISSION_DENIED) {
+        console.warn(
```

Comment:
Can we make sure, we only log this error once, in this case , in case of PERMISSION_DENIED , first a WARN message will be logged and then the the "ERROR" message. 
Also this will be logged for each batch if I am not wrong. 

A lot of our customers would be getting this Permission denied error, as they might not have required permissions as soon as they upgrade their client library, So we need to make sure that we don't log it too many times. Should not be more than 1 per minute(export interval).

Reply:
`Promise.all` consolidates the batch promises and the catch block is only executed once on error.  So each time the `export` is called there will be a one warning and and one error (if the user does not have permissions).  When we get to making use of the exporter I expect that we'll also use `PeriodicExportingMetricReader`, and we'll likely use a default of 60 seconds.